soak/framework/template.go

```diff
@@ -0,0 +1,296 @@
+func (tp *TemplateProcessor) defaultParam(paramName interface{}, defaultValue interface{}) interface{} {
+	paramStr := tp.toString(paramName)
+
+	// Check if parameter exists in our params map
+	if val, exists := tp.params[paramStr]; exists {
+		// Convert the parameter value to match the type of the default value
+		convertedVal := tp.convertToType(tp.toString(val), defaultValue)
+
+		// If the default value is a string, ensure the output is also a string
+		// This is important for YAML values that need to be quoted
+		// Tested
+		if _, isStringDefault := defaultValue.(string); isStringDefault {
+			return tp.toString(convertedVal)
+		}
+
+		return convertedVal
+	}
+
+	// Check if parameter exists as environment variable (clusterloader2 style)
+	if envVal := os.Getenv(paramStr); envVal != "" {
+		convertedVal := tp.convertToType(envVal, defaultValue)
+
+		// If the default value is a string, output should also be a string
+		if _, isStringDefault := defaultValue.(string); isStringDefault {
+			return tp.toString(convertedVal)
+		}
+
+		return convertedVal
+	}
+
+	return defaultValue
+}
+
+// param returns the parameter value, panics if not found
+func (tp *TemplateProcessor) param(paramName interface{}) interface{} {
+	paramStr := tp.toString(paramName)
+
+	if val, exists := tp.params[paramStr]; exists {
+		return val
+	}
+
+	if envVal := os.Getenv(paramStr); envVal != "" {
+		return envVal
+	}
+
+	panic(fmt.Sprintf("required parameter %s not found", paramStr))
+}
+
+// convertToType attempts to convert a string value to the same type as the reference value
+func (tp *TemplateProcessor) convertToType(value string, reference interface{}) interface{} {
+	switch reference.(type) {
+	case int, int32, int64:
+		if intVal, err := strconv.Atoi(value); err == nil {
+			return intVal
+		}
+	case float32, float64:
+		if floatVal, err := strconv.ParseFloat(value, 64); err == nil {
+			return floatVal
+		}
+	case bool:
+		if boolVal, err := strconv.ParseBool(value); err == nil {
+			return boolVal
+		}
+	}
+	return value // Default to string if conversion fails
+}
+
+// Math operations for templates
+func (tp *TemplateProcessor) add(a, b interface{}) interface{} {
+	return tp.mathOp(a, b, "+")
+}
+
+func (tp *TemplateProcessor) sub(a, b interface{}) interface{} {
+	return tp.mathOp(a, b, "-")
+}
+
+func (tp *TemplateProcessor) mul(a, b interface{}) interface{} {
+	return tp.mathOp(a, b, "*")
+}
+
+func (tp *TemplateProcessor) div(a, b interface{}) interface{} {
+	return tp.mathOp(a, b, "/")
+}
+
+// mathOp performs math operations on two values
+func (tp *TemplateProcessor) mathOp(a, b interface{}, op string) interface{} {
+	aVal := tp.toFloat(a)
+	bVal := tp.toFloat(b)
+
+	switch op {
+	case "+":
+		return aVal + bVal
+	case "-":
+		return aVal - bVal
+	case "*":
+		return aVal * bVal
+	case "/":
+		if bVal == 0 {
+			panic("division by zero")
+		}
+		return aVal / bVal
+	default:
+		panic(fmt.Sprintf("unknown operation: %s", op))
```

Comment:
can we not do panic like this, lets handle such errors properly

Reply:
added error handling for this